libp2p/connmanager.nim

```diff
@@ -0,0 +1,276 @@
+## Nim-LibP2P
+## Copyright (c) 2020 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+import tables, sequtils, sets
+import chronos, chronicles, metrics
+import peerinfo,
+       stream/connection,
+       muxers/muxer
+
+declareGauge(libp2p_peers, "total connected peers")
+
+const MaxConnectionsPerPeer = 5
+
+type
+  TooManyConnections* = object of CatchableError
+
+  MuxerHolder = object
+    muxer: Muxer
+    handle: Future[void]
+
+  ConnManager* = ref object of RootObj
+    # NOTE: don't change to PeerInfo here
+    # the reference semantics on the PeerInfo
+    # object itself make it succeptible to
+    # copies and mangling by unrelated code.
+    conns: Table[PeerID, HashSet[Connection]]
+    muxed: Table[Connection, MuxerHolder]
+    cleanUpLock: Table[PeerInfo, AsyncLock]
+    maxConns: int
+
+proc newTooManyConnections(): ref TooManyConnections {.inline.} =
+  result = newException(TooManyConnections, "too many connections for peer")
+
+proc init*(C: type ConnManager,
+           maxConnsPerPeer: int = MaxConnectionsPerPeer): ConnManager =
+  C(maxConns: maxConnsPerPeer,
+    conns: initTable[PeerID, HashSet[Connection]](),
+    muxed: initTable[Connection, MuxerHolder]())
+
+proc contains*(c: ConnManager, conn: Connection): bool =
+  ## checks if a connection is being tracked by the
+  ## connection manager
+  ##
+
+  if isNil(conn):
+    return
+
+  if isNil(conn.peerInfo):
+    return
+
+  if conn.peerInfo.peerId notin c.conns:
+    return
+
+  return conn in c.conns[conn.peerInfo.peerId]
```

Comment:
I would really avoid mixing weird nim return rules.
I would do like I refactored here:
https://github.com/status-im/nim-libp2p/blob/gossip-one-one/libp2p/peerinfo.nim#L142-L156

This is not really just about syntax but also helps the compiler predict branches, and most of all no branch left behind.

Reply:
I really dislike the no return on the other hand, it isn't explicit and the dangling values make code complicated to read and understand, not point in trusting the compiler if I can't follow the code. It isn't bad in this snippet, but it get bad in more involved ones.

Reply:
That's a cosmetic issue tho, if we have to take that into account there are so many that a I would change 🤣, my suggestion is to consider all branches and wether you put `return` or not does not matter, just always put a value, if you omit the keyword the compiler will analyze and error quite reliably as bonus.

Reply:
OK, I need to try it out more consistently. But, I understand the optimizations at play and those same optimizations are applied on return; as is an optimization to rely on the type's default value to avoid extra branches, that to me is more explicit and cleaner. 

My beef with this style is precisely style, this gets muddy really quickly and it is not immediately obvious to nim newcomers whats going on here. I always prefer explicit and "do what is expected" styles over magic optimizations and clever code - the simpler and dumber it is the easier it is to interpret by a human.